### DISCUSSION TITLE
What events trigger up & down arrows?  CEF based app not triggering them

### DISCUSSION THREAD

**Discussion by HarmlessDave** [nvaccess/nvda#12756] [Q&A]
Our app uses the Chromium Embedded Framework.  If we use CEF's older, lower-performance message handling scheme where our app handles the Windows message loop then the hotkeys for NVDA work properly.  If we switch to the higher performance "multi-threaded-mode" for CEF where it runs the message loop then many keys still work like a-z,0-9 but the arrow hot keys stop having any effect in NVDA.

I've tried forwarding WM_KEYDOWN and WM_KEYUP messages from the CEF browser window to the our mainframe that is hosting it, but that had no effect.  (There are no WM_CHAR messages for arrow keys so I can't send those.)

Is NVDA watching for some other event(s) in our main window that I could generate, or is there some message I can send directly to an NVDA window?

I've downloaded the NVDA source but it's taking a while to work through it to try to find the code related to the up and down arrows.

Edit: I see in the source that NVDA sets a Windows low-level keyboard hook (WH_KEYBOARD_LL).  

To see if our app was somehow no longer generating arrow key KEYUP and KEYDOWN events I created a simple logging hook DLL and loaded it into a second applet so that it would be attached to our app.  My logging DLL did get both up-arrow and down-arrow KEYUP and KEYDOWN events.

So I'm wondering why the NVDA code ignores these events from its hook.  Any suggestions would be helpful.

**Comment by seanbudd** [comment]
Hi @HarmlessDave,

Apologies for the slow response.
Are you expecting NVDA to receive [Window Messages](https://docs.microsoft.com/en-us/windows/win32/learnwin32/window-messages) sent from another window?
What effect in NVDA are you expecting?
Have you tried this with another screen reader such as Narrator?
Would you be able to provide some sample code and direct steps to reproduce?

**Reply by HarmlessDave** [reply]
Hi,

I only tried sending messages to NVDA's window to see if it would help, before finding the code where NVDA is using a keyboard hook.  Having the hook intercept and process arrow keys would be fine, it just stops working for us for some reason.

In CEF's single-threaded mode the keyboard hook set by NVDA works as expected and the arrow keys handle navigation for the reader.  In CEF's multi-threaded mode NVDA should still be getting the arrow keys (based on testing with my own keyboard hook) but instead NVDA is just passing the arrow keys on to the browser window for scrolling instead of moving the NVDA reading cursor to the next or previous line.

I'm wondering if it is because the window for the message loop switches from the main frame of our application to a hidden window created by the CEF framework, but I haven't dug deeply enough into the NVDA source to see if that makes sense.

For now we are hacking around this by switching to the low-performance single-threaded mode when we see NVDA in the process list.

This is a commercial application so I can't provide the real source code.  I guess this will need to stay unresolved until we have time to work up a Stack Overflow style minimal, complete and verifiable example (mcve).

**Comment by lukaszgo1** [comment]
It would be great to verify if NVDA indeed ignores the keys (less likely( or if it just cannot switch to browse mode (more likely). Assuming that you have an edit field in your application you can tab into it, type some text, try navigating with arrows and check if characters are announced. If they are then NVDA gets arrow keys but since it stays in focus mode arrows scroll the content rather than navigate with the virtual caret. If my hypothesis is correct then this is already reported as #8553

**Reply by HarmlessDave** [reply]
In both single-threaded and multi-threaded mode in an edit field (login page, user and password) the characters are read one-by-one as they are typed and the current full string is read when I pause typing.

In single threaded mode it also reads the full string again when I use the up and down arrow keys to move to the start or end of the field.  In multi-threaded mode the edit cursor moves to the start or end but nothing is read.

In single-threaded outside of an edit field the up and down move the reader forward or back, in multi-threaded the arrow keys just scroll the page.
